go/service/chat.go

```diff
@@ -4,38 +4,301 @@
+	if !ok {
+		return keybase1.Message{}, libkb.DecryptOpenError{}
+	}
+
+	if err := json.Unmarshal(plaintext, &unboxed.MessagePlaintext); err != nil {
+		return keybase1.Message{}, err
+	}
+
+	return unboxed, nil
+
+}
+
+// boxMessage encrypts a keybase1.MessagePlaintext into a chat1.MessageBoxed.  It
+// finds the most recent key for the TLF.
+func (h *chatLocalHandler) boxMessage(ctx context.Context, msg keybase1.MessagePlaintext) (chat1.MessageBoxed, error) {
+	tlfName := msg.ClientHeader.TlfName
+	keys, err := h.tlfh.CryptKeys(ctx, tlfName)
+	if err != nil {
+		return chat1.MessageBoxed{}, err
+	}
+
+	var recentKey *keybase1.CryptKey
+	for _, key := range keys.CryptKeys {
+		if recentKey == nil || key.KeyGeneration > recentKey.KeyGeneration {
+			recentKey = &key
+		}
+	}
+
+	if recentKey == nil {
+		return chat1.MessageBoxed{}, fmt.Errorf("no key found for tlf %q", tlfName)
+	}
+
+	return h.boxMessageWithKey(msg, recentKey)
+}
+
+// boxMessageWithKey encrypts a keybase1.MessagePlaintext into a chat1.MessageBoxed
+// given a keybase1.CryptKey.
+func (h *chatLocalHandler) boxMessageWithKey(msg keybase1.MessagePlaintext, key *keybase1.CryptKey) (chat1.MessageBoxed, error) {
+	s, err := json.Marshal(msg)
+	if err != nil {
+		return chat1.MessageBoxed{}, err
+	}
+
+	var nonce [libkb.NaclDHNonceSize]byte
+	if _, err := rand.Read(nonce[:]); err != nil {
+		return chat1.MessageBoxed{}, err
+	}
+
+	sealed := secretbox.Seal(nil, []byte(s), &nonce, ((*[32]byte)(&key.Key)))
+
+	boxed := chat1.MessageBoxed{
+		ClientHeader: msg.ClientHeader,
+		BodyCiphertext: chat1.EncryptedData{
+			V: 1,
+			E: sealed,
+			N: nonce[:],
+		},
+		KeyGeneration: key.KeyGeneration,
+	}
+
+	return boxed, nil
 }
 
-func (h *chatLocalHandler) PostLocal(context.Context, keybase1.PostLocalArg) error {
+// signMessageBoxed signs the header and encrypted body of a chat1.MessageBoxed
+// with the NaclSigningKeyPair.
+func (h *chatLocalHandler) signMessageBoxed(msg *chat1.MessageBoxed, kp libkb.NaclSigningKeyPair) error {
+	header, err := h.sign(msg.ClientHeader, kp)
+	if err != nil {
+		return err
+	}
+	msg.HeaderSignature = header
+
+	body, err := h.sign(msg.BodyCiphertext, kp)
+	if err != nil {
+		return err
+	}
+	msg.BodySignature = body
+
 	return nil
 }
+
+// sign signs data with a NaclSigningKeyPair, returning a chat1.SignatureInfo.
+func (h *chatLocalHandler) sign(data interface{}, kp libkb.NaclSigningKeyPair) (chat1.SignatureInfo, error) {
+	encoded, err := json.Marshal(data)
+	if err != nil {
+		return chat1.SignatureInfo{}, err
+	}
+
+	sig := *kp.Private.Sign(encoded)
+
+	info := chat1.SignatureInfo{
+		V: 1,
+		S: sig[:],
+		K: kp.Public[:],
+	}
+
+	return info, nil
+}
+
+// keyFinder remembers results from previous calls to CryptKeys().
+// It is not intended to be used by multiple concurrent goroutines
+// or held onto for very long, just to remember the keys while
+// unboxing a thread of messages.
+type keyFinder struct {
+	keys map[string]keybase1.TLFCryptKeys
```

Comment:
it's only used as a local variable for one ThreadViewBoxed -> ThreadView conversion, then it goes away.